### PR TITLE
Fix concurrent workspace limit

### LINKED ISSUE #9410: Max. parallel workspace limit is not correctly enforced
We initialize new instances with `phase: "unknown"` [here](https://github.com/gitpod-io/gitpod/blob/ed30d9696fd1c0ca5775a2d44d6b4a9e5709ba63/components/server/src/workspace/workspace-starter.ts#L690). When enforcing the max parallel workspace limit we filter that phase out, because we do not want to block people whose workspaces are stuck on nodes that got lost due to a data center went down.

We should strive to fix this by using `"preparing"` as initial state.

Slightly related, but not to confused with: https://github.com/gitpod-io/gitpod/issues/8830

### PR DESCRIPTION
**Pull request by andrew-farries**
## Description

Change the initial phase of a workspace from `unknown` to `preparing` to avoid filtering them out when calculating a user's concurrent workspace limit.

Builds on #9453 to allow us to distinguish between workspaces waiting for an image build and workspaces that may not need one.

## Related Issue(s)

Fixes #9410

## How to test

Smoke test various IDEs:

- [x] In-browser vscode
- [x] Desktop vscode
- [x] Goland (via Jetbrains Gateway)
- [x] local companion app

## Release Notes

```release-note
Correctly enforce the parallel workspace limit
```

## Documentation

None

 - [x] /werft with-vm=true

### CODE REVIEW

**Comment by meysholdt** [comment]
hey, can you please rebase your branch before you continue working with your preview env? The old base contains a ws-daemon that [logs expensively](https://github.com/gitpod-io/gitpod/issues/9530).

To save costs, I'll manually remove the old preview env of this branch. Apologies for the inconvenience.

**Comment by csweichel** [comment]
@andrew-farries @geropl how do we move forward with this? (post-rebase that is :) )

**Comment by andrew-farries** [comment]
I'm struggling to test this (and the branch on which it's based - #9453) because image builds aren't working for me in preview environments. It needs a a run to check that we haven't broken anything in the transition between workspace phases with the addition of the new 'building' phase.

**Comment by geropl** [comment]
> @andrew-farries @geropl how do we move forward with this? (post-rebase that is :) )

@csweichel  The first PR of the series (the types and backwards-compatible handling) is merged: #9453

Before moving forward with this PR we now have to make sure all clients handle the new "phase".

**Comment by andrew-farries** [comment]
rebased to resolve merge conflicts and taking this out of 'draft'.

**Comment by geropl** [comment]
@andrew-farries Moving back to draft because we have to clarify with @gitpod-io/engineering-ide what they have to do before we can merge this one. :+1:

**Comment by akosyakov** [comment]
It seems that we don't have incompatible changes with it:
- in JB GW it seems to be alright already: https://github.com/gitpod-io/gitpod/blob/c6fec61ebd4f180e87fd0b3c63c3205f1c55e224/components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt#L165-L172
- in local app it seems to be backward compatible, we only care there about `running`, `stopping` and `stopped`: https://github.com/gitpod-io/gitpod/blob/c6fec61ebd4f180e87fd0b3c63c3205f1c55e224/components/local-app/pkg/bastion/bastion.go#L302
- in supervisor frontend it seems to be the same: https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/frontend/src/index.ts
- in supervisor backend it should be the same
- I don't think that VS Code extension for Gitpod actually care about phases.

That's said it is appreciated if you go and smoke testing of IDEs in regards of starting, stopping and restarting.

**Comment by geropl** [comment]
@andrew-farries Could you take over the testing of various IDEs? :pray:

**Comment by andrew-farries** [comment]
I finally have some time again to look at this. I've updated the 'how to test' section with what we need to do and checked that most things still work.

@akosyakov How do I configure the Jetbrains gateway to be able to connect to workspaces in a preview environment?

**Comment by akosyakov** [comment]
@andrew-farries go to preferences of your prev env select JB product, start a workspace and follow the flow

File: components/server/src/workspace/workspace-starter.ts [hunk 1]
```diff
@@ -989,7 +989,7 @@ export class WorkspaceStarter {
             // Update workspace instance to tell the world we're building an image
             const workspaceImage = result.ref;
             const status: WorkspaceInstanceStatus = result.actuallyNeedsBuild
-                ? { ...instance.status, phase: "preparing" }
+                ? { ...instance.status, phase: "building" }
```
Comment (AlexTugarev) [inline components/server/src/workspace/workspace-starter.ts:992]: 👍🏻

File: components/ws-manager-bridge/src/config.ts [hunk 2]
```diff
@@ -33,6 +33,7 @@ export interface Configuration {
     timeouts: {
         metaInstanceCheckIntervalSeconds: number;
         preparingPhaseSeconds: number;
+        buildingPhaseSeconds: number;
```
Comment (AlexTugarev) [inline components/ws-manager-bridge/src/config.ts:36]: how do we distribute config changes? do we need to have default values for compatibility?
what's the plan for self-hosted?

**Comment by AlexTugarev** [comment]
I'm testing now...

File: components/ws-manager-bridge/src/config.ts (hunk 2 above)
Comment (AlexTugarev) [reply components/ws-manager-bridge/src/config.ts:36]: @andrew-farries, I believe this needs to be reflected in configmap.go, like the other timeouts.

Comment (andrew-farries) [reply components/ws-manager-bridge/src/config.ts:36]: Done in a0e4ea775123ea0476c2c43814f189f6074d183e

**Pull request merged**